enhancements/cluster-api/cluster-autoscaler-integration-with-openshift-cluster-api.md

```diff
@@ -0,0 +1,751 @@
+metadata:
+  name: "worker-somezone-1"
+  namespace: "openshift-machine-api"
+spec:
+  minReplicas: 1
+  maxReplicas: 12
+  scaleTargetRef:
+    apiVersion: machine.openshift.io/v1beta1
+    kind: MachineSet
+    name: example-cluster-worker-somezone-1
+```
+
+```yaml
+apiVersion: "autoscaling.openshift.io/v1beta1"
+kind: "MachineAutoscaler"
+metadata:
+  name: "worker-somezone-2"
+  namespace: "openshift-machine-api"
+spec:
+  minReplicas: 1
+  maxReplicas: 12
+  scaleTargetRef:
+    apiVersion: cluster.x-k8s.io/v1beta1
+    kind: MachineSet
+    name: example-cluster-worker-somezone-2
+```
+
+Note that the MachineAutoscaler named "worker-somezone-1" is targeting a Machine API
+MachineSet while "worker-somezone-2" is targeting a Cluster API MachineSet. The
+Cluster Autoscaler Operator will know by the `apiVersion` field whether to look
+for the resource in the `openshift-machine-api` or `openshift-cluster-api` namespace
+respectively.
+
+To support existing user experiences and workflows, the user might also have created
+the "worker-somezone-2" MachineAutoscaler using the Machine API reference as follows:
+
+```yaml
+apiVersion: "autoscaling.openshift.io/v1beta1"
+kind: "MachineAutoscaler"
+metadata:
+  name: "worker-somezone-2"
+  namespace: "openshift-machine-api"
+spec:
+  minReplicas: 1
+  maxReplicas: 12
+  scaleTargetRef:
+    apiVersion: machine.openshift.io/v1beta1
+    kind: MachineSet
+    name: example-cluster-worker-somezone-2
+```
+
+The Cluster Autoscaler Operator will update the Machine API MachineSet resource, and
+then the MachineSet sync controller will sync the change to the Cluster API MachineSet
+resource. The sync controller will use the managed fields (i.e. `.metadata.managedFields`)
+of the specified MachineSet to determine if the Cluster Autoscaler Operator made
+changes to the annotations, and then replicate those appropriately. In this manner,
+a user might specify an authoritative or non-authoritative MachineSet in the
+`scaleTargetRef` and the sync controller will be able to properly mirror the
+changes by detecting that the Cluster Autoscaler Operater has authored the changes.
+
+It will be considered an error, represented by a condition on the MachineAutoscaler resource,
+if a user creates two MachineAutoscalers referencing both the Machine API
+and Cluster API MachineSets for the same physical instances. In this scenario, the second
+MachineAutoscaler resource will gain an error condition regardless of which MachineSet is
+authoritative.
+
+The Cluster Autoscaler Operator will always update the authoritative MachineSet resource.
+If a user specifies a non-authoritative MachineSet as the `scaleTargetRef` of a
+MachineAutoscaler, the Cluster Autoscaler Operator will use the information on the
+MachineSet to determine which resource is authoritative and then update that resource.
+Through the MachineSet sync controller, the non-authoritative resource will be updated
+with the new information.
```

Comment:
Maybe I'm getting confused, but we have two components here.

One is CAO, which we _could_ make smart enough to understand the authoritative API?

And the other is KAS itself, which will always write to CAPI resources, right?

Reply:
correct